File: app/src/main/java/com/ishilab/transcriber/transcribe/WhisperEngine.kt
package com.ishilab.transcriber.transcribe

import android.util.Log
import com.ishilab.transcriber.whisper.WhisperLib

/**
 * whisper.cpp を用いたローカル文字起こしエンジン。
 *
 * @param modelPath 端末上の ggml モデルファイルへの絶対パス
 * @param language  認識言語（既定 "ja"）
 */
class WhisperEngine(
    private val modelPath: String,
    private val language: String = "ja"
) : TranscriptionEngine {

    private var contextPtr: Long = 0L

    override val isLoaded: Boolean
        get() = contextPtr != 0L

    // 実時間に追いつけるよう、使えるコアを多めに使う（上限8）。
    private val numThreads: Int
        get() = maxOf(2, minOf(8, Runtime.getRuntime().availableProcessors()))

    override fun load() {
        if (isLoaded) return
        contextPtr = WhisperLib.initContext(modelPath)
        check(contextPtr != 0L) { "whisper モデルの読み込みに失敗しました: $modelPath" }
        Log.i(TAG, "whisper loaded. system=${WhisperLib.getSystemInfo()} threads=$numThreads")
    }

    @Synchronized
    override fun transcribe(samples: FloatArray): String {
        check(isLoaded) { "engine not loaded" }
        val rc = WhisperLib.fullTranscribe(contextPtr, numThreads, language, samples)
        if (rc != 0) {
            Log.w(TAG, "fullTranscribe returned $rc")
            return ""
        }
        val count = WhisperLib.getTextSegmentCount(contextPtr)
        val sb = StringBuilder()
        for (i in 0 until count) {
            sb.append(WhisperLib.getTextSegment(contextPtr, i).trim())
        }
        return sb.toString().trim()
    }

    @Synchronized
    override fun release() {
        if (contextPtr != 0L) {
            WhisperLib.freeContext(contextPtr)
            contextPtr = 0L
        }
    }

    companion object {
        private const val TAG = "WhisperEngine"
    }
}

File: ios/Transcriber/Transcribe/WhisperEngine.swift
import Foundation

/// whisper.cpp を用いたローカル文字起こしエンジン。
/// Android 版は JNI (WhisperLib) 経由だったが、iOS ではブリッジングヘッダから
/// C API を直接呼ぶ。
final class WhisperEngine: TranscriptionEngine {

    private let modelPath: String
    private let language: String
    private var ctx: OpaquePointer?
    private let lock = NSLock()

    var isLoaded: Bool { ctx != nil }

    /// 実時間に追いつけるよう、使えるコアを多めに使う（上限8）。
    private var numThreads: Int32 {
        Int32(max(2, min(8, ProcessInfo.processInfo.processorCount)))
    }

    init(modelPath: String, language: String = "ja") {
        self.modelPath = modelPath
        self.language = language
    }

    func load() throws {
        if isLoaded { return }
        var cparams = whisper_context_default_params()
        cparams.use_gpu = false // Android と同じ CPU バックエンドのみ
        ctx = whisper_init_from_file_with_params(modelPath, cparams)
        guard ctx != nil else {
            throw NSError(domain: "WhisperEngine", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "whisper モデルの読み込みに失敗しました: \(modelPath)"])
        }
        NSLog("WhisperEngine: whisper loaded. system=%@ threads=%d",
              String(cString: whisper_print_system_info()), numThreads)
    }

    func transcribe(_ samples: [Float]) -> String {
        lock.lock(); defer { lock.unlock() }
        guard let ctx else { return "" }
        var params = whisper_full_default_params(WHISPER_SAMPLING_GREEDY)
        params.print_realtime = false
        params.print_progress = false
        params.print_timestamps = false
        params.print_special = false
        params.translate = false
        params.n_threads = numThreads
        params.offset_ms = 0
        params.no_context = true
        params.single_segment = false

        let rc: Int32 = language.withCString { lang in
            params.language = lang
            return samples.withUnsafeBufferPointer { buf in
                whisper_full(ctx, params, buf.baseAddress, Int32(buf.count))
            }
        }
        if rc != 0 {
            NSLog("WhisperEngine: whisper_full returned %d", rc)
            return ""
        }
        let count = whisper_full_n_segments(ctx)
        var out = ""
        for i in 0..<count {
            if let text = whisper_full_get_segment_text(ctx, i) {
                out += String(cString: text).trimmingCharacters(in: .whitespaces)
            }
        }
        return out.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func release() {
        lock.lock(); defer { lock.unlock() }
        if let c = ctx {
            whisper_free(c)
            ctx = nil
        }
    }
}
